### PR TITLE
Fix SerializerPropertyMetadataFactory

### PR DESCRIPTION
**Pull request by teohhanhui**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | N/A

* Set readable / writable based on serializer groups of target property
* Set readableLink / writableLink based on serializer groups of properties on related resource

### CODE REVIEW

**Comment by teohhanhui** [comment]
The previous behaviour was incorrect. We'll need to add tests to prevent regression.

**Comment by dunglas** [comment]
:+1:, fell free to merge it when you'll have some time to add tests.

**Comment by teohhanhui** [comment]
Now with more fixes.

ping @api-platform/core-team

File: tests/Metadata/Property/Factory/SerializerPropertyMetadataFactoryTest.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                ],
+                'denormalization_context' => [
+                    'groups' => ['dummy_write'],
+                ],
+            ]);
+        $resourceMetadataFactoryProphecy->create(Dummy::class)->willReturn($dummyResourceMetadata)->shouldBeCalled();
+        $resourceMetadataFactoryProphecy->create(RelatedDummy::class)->willReturn(new ResourceMetadata())->shouldBeCalled();
+        $resourceMetadataFactory = $resourceMetadataFactoryProphecy->reveal();
+
+        $serializerClassMetadataFactoryProphecy = $this->prophesize(SerializerClassMetadataFactoryInterface::class);
```
Comment (teohhanhui) [inline tests/Metadata/Property/Factory/SerializerPropertyMetadataFactoryTest.php]: This whole unreadable chunk here is due to the lack of fluent interface in Symfony Serializer :smile:

File: src/Metadata/Property/Factory/SerializerPropertyMetadataFactory.php [hunk 2]
```diff
@@ -46,9 +47,38 @@ public function create(string $resourceClass, string $property, array $options =
...
+            }
+        }
+
+        /*
+         * Set readable/writable based on matching normalization/denormalization groups,
+         * but never reset a false value as it could be unreadable/unwritable for other reasons
+         *
+         * If normalization/denormalization groups are not specified, the property is implicitly readable/writable
+         */
+        if (false !== $propertyMetadata->isReadable()) {
```
Comment (theofidry) [inline src/Metadata/Property/Factory/SerializerPropertyMetadataFactory.php]: coud L67-75 be moved into a private `::getPropertyMetadata()` method? Then those comments could go in the method PHPDoc. That would increase readability IMO

File: tests/Metadata/Property/Factory/SerializerPropertyMetadataFactoryTest.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
+<?php
```
Comment (theofidry) [inline tests/Metadata/Property/Factory/SerializerPropertyMetadataFactoryTest.php]: :+1:

File: tests/Metadata/Property/Factory/SerializerPropertyMetadataFactoryTest.php (hunk 1 above)
Comment (theofidry) [inline tests/Metadata/Property/Factory/SerializerPropertyMetadataFactoryTest.php]: no blaming ya for 'dat :p

**Comment by theofidry** [comment]
Looks good to me, I would just prefer a more expressive and less declarative code to make it more readable but maybe @dunglas will say it's just me mixing with the wrong kind :sweat_smile:

**Comment by teohhanhui** [comment]
@theofidry
> I would just prefer a more expressive and less declarative code to make it more readable

What do you mean?

**Comment by teohhanhui** [comment]
I can think of maybe `transformReadWrite` and `transformLinkStatus` private functions. Lol...

What do you think?

**Comment by theofidry** [comment]
I mean in places where you have:

```php
$myProp = ..
// do some check or something to update $myProp
```

I find more readable to have just:

```php
$myProp = myProp(...);
```

and move everything in a separate function. In short "Tell, don't ask": [more detailed article here](https://robots.thoughtbot.com/tell-dont-ask).

**Comment by teohhanhui** [comment]
Better now? :smile:

**Comment by theofidry** [comment]
:ok_hand:

**Comment by teohhanhui** [comment]
Further refactoring. Should make Scrutinizer happy.

**Comment by dunglas** [comment]
:+1:

**Pull request merged**